### PR TITLE
fix chain multiplier modifiers not benefiting from spell power or healing power

### LINKED ISSUE #915: Chain Lightning Bonus not benefitting from Spell Power
Similar to my Improved Chain Heal (ICH) (T2 3 set bonus report)

http://db.vanillagaming.org/?spell=26123 is the spell from the Gift of the Gathering Storm set.

Default decay of chain lightning is 0.7x, 0.49 for the second bounce. CLB is a 5% bonus (0.75 0.5625).

This may not be conclusive proof (harder to test with a small bonus) but I believe it is enough to warrant investigation into the code. I tested it similarly to ICH by manually adding the aura to character. Tested it on Bropack v7. This time I included criticals (divided by 1.5) and I used resists (added damage + resisted amount). I tested it on the imp packs in molten core.

T1, T2, T3 = damage to target 1, 2, 3

| Test  | Trials | Avg (T2/T1) | Avg (T3/T2) | Avg (T3/T1) | AVG T1 | AVG T2 | AVG T3 | 
| ------------- | ------------- | ------------- | ------------- |------------- |------------- |------------- |------------- |
| 0+Spell | 51 | 0.7079 | 0.6896 | 0.4882 | 529.3987 | 374.7647 |  258.4510 |
| CLB 0+Spell | 88 | 0.7317 | 0.7336 | 0.5368 | 535.9432 | 392.1667 | 287.7083 |
| 640+Spell | 72 | 0.7013 | 0.6994 | 0.4905 | 991.9398 | 695.6852 | 486.5370 |
| CLB 640+Spell | 65 | 0.7134 | 0.7207 | 0.5141 | 996.1077 | 710.5897 | 512.1026 |

Difference between AVG T3 for 0+Spell = 29.26, 640+spell = 25.56

combat logs are uploaded here: https://gist.github.com/zelazrgn

Side note: possible bug is that if you get a full resist on T1, the damage done on T2 and T3 is greatly increased. It seems that maybe T2 and T3 may get the bonus of T1 and T2 if the original target fully resists, I didn't purposely test this but it was a significant boost (+100 damge on T3).

I would also like to re-open https://github.com/LightsHope/server/issues/894 on the following grounds:

1. The earliest source on the benefits of the T2 set bonus indicates it does benefit from +healing. Nothing indicates that wasn't the case in 1.12.1.
2. These spells are very unique and behavior of more traditional talents cannot be assumed. It seems unreasonable that they would adjust it to benefit from +healing only to then nerf it from 30% down to 5% in 2.4.2.
3. Behavior makes more sense from end user, they maintain a consistent ratio from T1/T2/T3 as you add HP.
4. Both these spells (ICH, CLB) are nearly identical looking at QSW.
5. Chain Healing Wave (T1 bonus) does benefit from +healing in this code base.

Last 2 are suggesting that the behavior of these spells should probably be consistent.

### PR DESCRIPTION
**Pull request by zelazrgn**
fixes
#915
#894

I copied the ApplySpellMod from Spell::DoSpellHitOnUnit
https://github.com/LightsHope/server/blob/4d33e2450fad4a314b69e3940ed13f9f4579c08b/src/game/Spells/Spell.cpp#L1632

into CalculateCustomCoefficient. ApplySpellMod has a side effect of dropping mod charges but that is not an issue since these spells do not have charges.

http://vanillawowdb.com/?spell=21899
http://vanillawowdb.com/?spell=26123

### CODE REVIEW

**Comment by zelazrgn** [comment]
My issues are mostly results from testing these two spells so I will summarize my evidence and reason for this change here.

Improved Chain Heal is the 3 set bonus from Ten Storms
Increases the amount healed by Chain Heal to targets beyond the first by 30%.
http://vanillawowdb.com/?spell=21899


Gift of the Gathering Storm Chain Lightning Bonus is from the Gift of the Gathering set.
Increases the chain target damage multiplier of your Chain Lightning spell by 5%.
http://vanillawowdb.com/?spell=26123

These are the only two spells from Vanilla that shared the spell mod "SPELLMOD_EFFECT_PAST_FIRST".
http://web.archive.org/web/20060313052257/http://www.thottbot.com:80/?e=Apply%20Aura%3A%20Add%20%25%20Modifier%20%28Effect%20Past%20First%3F%29
(excluding 16183 which I imagine was never used)

They both use different language but they are nearly identical in the spell dbc file. They are treated the exact same way in the source code and the current behavior is that SPELLMOD_EFFECT_PAST_FIRST effects are ignored when calculating spell power/healing power coefficients.

Ignoring spell power/healing power, these effects modify the multiplier so that for Chain heal, it is increased from 0.5 to 0.65 (0.5 * 1.3), and Chain Lightning 0.70 to 0.735 (0.70 * 1.05).

The earliest source I have found is dated 1/7/2008 (Patch 2.3) for the section on tier 2 and references two other sources for the calculations. https://web.archive.org/web/20080302191341/http://elitistjerks.com:80/f31/t19181-shaman_how_heal_like_pro/

The only evidence against this change is that +healing did not benefit from percent increase talents (such as improved renew) however these spell effects are unique SPELLMOD_EFFECT_PAST_FIRST and, at least in this source code, applied in a separate way. The end result is that these chain spells will maintain a consistent ratio between bounces regardless of the amount of +sp/hp. There is no reason to believe the current behavior is correct for the Gift of the Gathering Storm bonus.

File: src/game/Spells/SpellMgr.cpp [hunk 1]
```diff
@@ -328,8 +328,16 @@ float CalculateCustomCoefficient(SpellEntry const *spellProto, Unit const* caste
                 // Chain Lightning / Chain Heal / Healing Wave (T1 8/8 bonus)
                 if (spellProto->IsFitToFamilyMask(UI64LIT(0x00000000142)))
                 {
+                    float multiplier = spellProto->DmgMultiplier[0];
+
```
Comment (thepparker) [inline src/game/Spells/SpellMgr.cpp]: Wrap this inside an if (spell->GetTargetNum() > 1)

**Comment by thepparker** [comment]
Given that we already apply power bonuses to chain bounces depending on the target number, it makes sense that the set bonuses would apply to the bonus coefficient as well. Merge is pending change in comments.

File: src/game/Spells/SpellMgr.cpp (hunk 1 above)
Comment (thepparker) [reply src/game/Spells/SpellMgr.cpp]: by this, I mean the spellmod application

**Comment by thepparker** [comment]
Yea, I know. I meant wrap the `if (Player* player =...`. Just a minor optimization, so we skip it much like the coeff multiplication on the first target

**Comment by zelazrgn** [comment]
all set now :)

**Pull request merged**